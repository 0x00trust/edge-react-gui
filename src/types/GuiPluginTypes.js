// @flow

import { asArray, asEither, asNumber, asObject, asOptional, asString } from 'cleaners'

import { type Permission } from '../reducers/PermissionsReducer.js'

/**
 * A unique WebView-based plugin.
 *
 * This contains just the information needed to launch the plugin,
 * either from the plugin list or from a deep link.
 *
 * The plugin list uses additional information for its sorting & filtering,
 * which lives in other data structures.
 */
export type GuiPlugin = {
  // This pluginId should be short and use dashes to separate words,
  // since it appears in deep links like `edge://plugin/moonpay-buy`:
  pluginId: string,

  // The storage location to make available in the `EdgeProvider`.
  // Also used for conversion tracking:
  storeId: string,

  // The URI to show in the WebView.
  // Both the plugin list & deep links can add stuff to the end of this:
  baseUri: string,

  // Scene title to display when inside the plugin:
  displayName: string,

  // Use the Airbitz-based API when set, instead of EdgeProvider:
  isLegacy?: true,

  // The WebView won't navigate to hostnames outside of this list:
  originWhitelist?: string[],

  // Device permissions to acquire before launching the plugin:
  permissions?: Permission[]
}

/**
 * A row in the plugin list scene, after being distilled down from JSON.
 */
export type GuiPluginRow = {
  pluginId: string,
  addOnUrl: string,

  title: string,
  description: string,
  partnerIconPath?: string,
  paymentTypeLogoKey?: string,
  paymentTypes: string[],
  cryptoCodes: string[]
}

/**
 * The plugin list scene stores its data in JSON files,
 * which have an array of these rows mixed with strings for comments.
 */
const asGuiPluginJsonRow = asObject({
  // A unique string to identify this particular row:
  id: asString,

  // The plugin to display if we select this item:
  pluginId: asOptional(asString),

  // Optional suffix to add to plugin URI:
  addOnUrl: asOptional(asString),

  // List display options:
  title: asOptional(asString),
  description: asOptional(asString),
  partnerIconPath: asOptional(asString),
  paymentTypeLogoKey: asOptional(asString),
  paymentTypes: asOptional(asArray(asString)),
  cryptoCodes: asOptional(asArray(asString)),

  // Filtering & sorting:
  forCountries: asOptional(asArray(asString)),
  forPlatform: asOptional(asString),
  sortIndex: asOptional(asNumber)
})
export const asGuiPluginJson = asArray(asEither(asString, asGuiPluginJsonRow))
export type GuiPluginJson = $Call<typeof asGuiPluginJson, any>

/**
 * Helper function to turn a GuiPluginJson into a cooked list.
 * Call `asGuiPluginJson` to clean & validate the input file first.
 */
export function filterGuiPluginJson (cleanJson: GuiPluginJson, platform: string, countryCode: string): GuiPluginRow[] {
  // Filter and merge related rows:
  const mergedRows: { [id: string]: GuiPluginRow } = {}
  const sortIndexes: { [id: string]: number } = {}
  for (const row of cleanJson) {
    if (typeof row === 'string') continue

    // Filtering:
    const { id, forCountries, forPlatform, sortIndex } = row
    if (forCountries != null && forCountries.indexOf(countryCode) < 0) continue
    if (forPlatform != null && forPlatform !== platform) continue
    if (sortIndex != null) sortIndexes[id] = sortIndex

    // Defaults:
    if (sortIndexes[id] == null) sortIndexes[id] = 0
    if (mergedRows[id] == null) {
      mergedRows[id] = {
        pluginId: '',
        addOnUrl: '',
        title: '',
        description: '',
        paymentTypes: [],
        cryptoCodes: []
      }
    }

    // Merging:
    const merged = mergedRows[id]
    if (row.pluginId != null) merged.pluginId = row.pluginId
    if (row.addOnUrl != null) merged.addOnUrl = row.addOnUrl
    if (row.title != null) merged.title = row.title
    if (row.description != null) merged.description = row.description
    if (row.partnerIconPath != null) merged.partnerIconPath = row.partnerIconPath
    if (row.paymentTypeLogoKey != null) merged.paymentTypeLogoKey = row.paymentTypeLogoKey
    if (row.paymentTypes != null) merged.paymentTypes = row.paymentTypes
    if (row.cryptoCodes != null) merged.cryptoCodes = row.cryptoCodes
  }

  // Build the sorted output list, removing rows without pluginIds:
  return Object.keys(mergedRows)
    .filter(id => mergedRows[id].pluginId !== '')
    .sort((a, b) => sortIndexes[a] - sortIndexes[b])
    .map(id => mergedRows[id])
}
