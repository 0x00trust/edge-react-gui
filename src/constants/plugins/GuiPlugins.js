// @flow

import { Platform } from 'react-native'
import RNFS from 'react-native-fs'

import { type BuySellPlugin, type GuiPlugin } from '../../types/GuiPluginTypes.js'

const buyPlugins = require('./buyPluginList.json')
const sellPlugins = require('./sellPluginList.json')

const hostedUri = Platform.OS === 'android' ? 'file:///android_asset/plugins/' : `file:///${RNFS.MainBundlePath}/plugins/`

export const guiPlugins: { [pluginId: string]: GuiPlugin } = {
  libertyx: {
    pluginId: 'libertyx',
    storeId: 'com.libertyx',
    baseUri: 'https://libertyx.com/a/',
    displayName: 'LibertyX',
    originWhitelist: ['https://libertyx.com'],
    permissions: ['location']
  },
  moonpay: {
    pluginId: 'moonpay',
    storeId: 'io.moonpay.buy',
    baseUri: 'https://buy.moonpay.io?apiKey=pk_live_Y1vQHUgfppB4oMEZksB8DYNQAdA4sauy',
    displayName: 'MoonPay'
  },
  safello: {
    pluginId: 'safello',
    storeId: 'com.safello',
    baseUri: 'https://safello.com/edge/',
    displayName: 'Safello',
    originWhitelist: ['https://safello.com', 'https://app.safello.com', 'http://safello.com']
  },
  'safello-sell': {
    pluginId: 'safello-sell',
    storeId: 'com.safello',
    baseUri: 'https://app.safello.com/',
    displayName: 'Safello',
    originWhitelist: ['https://safello.com', 'https://app.safello.com', 'http://safello.com']
  },
  bitsofgold: {
    pluginId: 'bitsofgold',
    storeId: 'bitsofgold',
    baseUri: 'https://www.bitsofgold.co.il/order/',
    displayName: 'Bits of Gold',
    permissions: ['camera']
  },
  banxa: {
    pluginId: 'banxa',
    storeId: 'banxa',
    baseUri: 'https://edge.banxa.com',
    displayName: 'Banxa',
    permissions: ['camera']
  },
  simplex: {
    pluginId: 'simplex',
    storeId: 'co.edgesecure.simplex',
    baseUri: hostedUri + 'co.edgesecure.simplex/index.html',
    displayName: 'Simplex'
  },
  wyre: {
    pluginId: 'wyre',
    storeId: 'co.edgesecure.wyre',
    baseUri: hostedUri + 'co.edgesecure.wyre/index.html',
    displayName: 'Wyre',
    permissions: ['camera']
    // supportEmail: 'support@sendwyre.com'
  },
  bity: {
    pluginId: 'bity',
    storeId: 'com.bity',
    baseUri: hostedUri + 'com.bity/index.html',
    displayName: 'Bity'
    // supportÃŸEmail: 'support@bity.com'
  },
  bitrefill: {
    pluginId: 'bitrefill',
    storeId: 'co.edgesecure.bitrefill',
    baseUri: hostedUri + 'co.edgesecure.bitrefill/index.html',
    displayName: 'Bitrefill',
    isLegacy: true
  },
  cred: {
    pluginId: 'cred',
    storeId: 'cred',
    baseUri: 'https://earn.mycred.io/edge',
    displayName: 'Cred',
    permissions: ['camera']
  },
  custom: {
    pluginId: 'custom',
    storeId: 'custom',
    baseUri: '',
    displayName: 'Custom Plugin',
    permissions: ['camera', 'location']
  }
}

export const devPlugin: BuySellPlugin = {
  pluginId: 'custom',
  id: '',
  priority: 99,
  paymentType: '',
  description: '',
  title: 'Custom Dev',
  paymentTypeLogoKey: 'credit',
  partnerIconPath: '',
  cryptoCodes: []
}

export const collapsePlugins = function (pluginsRaw: Array<string | BuySellPlugin>, platform: string, countryCode: string): Array<BuySellPlugin> {
  const collapsedById: { [id: string]: BuySellPlugin } = {}

  // Collapse all
  pluginsRaw.forEach(pluginObjOrString => {
    if (typeof pluginObjOrString !== 'string') {
      let objToMerge
      if (pluginObjOrString.forPlatform && pluginObjOrString.forPlatform === platform) {
        objToMerge = pluginObjOrString
      } else if (pluginObjOrString.forCountries && pluginObjOrString.forCountries.includes(countryCode)) {
        objToMerge = pluginObjOrString
      } else if (!pluginObjOrString.forPlatform && !pluginObjOrString.forCountries) {
        objToMerge = pluginObjOrString
      }
      const id = pluginObjOrString.id
      collapsedById[id] = { ...collapsedById[id], ...objToMerge }
    }
  })
  const pluginsAll: Array<BuySellPlugin> = Object.keys(collapsedById).map((id: string) => collapsedById[id])

  // Filter by countryCode
  const pluginsFilteredByCountryCode: Array<BuySellPlugin> = pluginsAll.filter((pluginObj: BuySellPlugin) => {
    if (pluginObj.countryCodes) {
      return pluginObj.countryCodes[countryCode]
    }
    return false
  })

  // Delete countryCodes
  return pluginsFilteredByCountryCode.map(pluginObj => {
    delete pluginObj.countryCodes
    return pluginObj
  })
}

export const getBuyPlugins = function (platform: string, countryCode: string): Array<BuySellPlugin> {
  return collapsePlugins(buyPlugins, platform, countryCode)
}

export const getSellPlugins = function (platform: string, countryCode: string): Array<BuySellPlugin> {
  return collapsePlugins(sellPlugins, platform, countryCode)
}
