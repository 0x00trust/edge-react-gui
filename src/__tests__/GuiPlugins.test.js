// @flow
/* globals describe it expect */

import { collapsePlugins, guiPlugins } from '../constants/plugins/GuiPlugins.js'
import { type BuySellPlugin } from '../types/GuiPluginTypes.js'

const buyPluginJson = require('../constants/plugins/buyPluginList.json')
const sellPluginJson = require('../constants/plugins/sellPluginList.json')

describe('Production plugin data', () => {
  it('Has accurate pluginId fields', () => {
    const pluginIds = Object.keys(guiPlugins)
    for (const pluginId of pluginIds) {
      expect(guiPlugins[pluginId].pluginId).toEqual(pluginId)
    }
  })

  it('Buy & sell plugins have real pluginId fields', () => {
    for (const row of [...buyPluginJson, ...sellPluginJson]) {
      if (typeof row === 'string') continue
      const { pluginId } = row
      if (pluginId != null) expect(guiPlugins[pluginId]).not.toBeUndefined()
    }
  })

  it('Buy plugins match snapshot on iOS + US', () => {
    expect(collapsePlugins(buyPluginJson, 'ios', 'US')).toMatchSnapshot()
  })

  it('Sell plugins match snapshot on iOS + US', () => {
    expect(collapsePlugins(sellPluginJson, 'ios', 'US')).toMatchSnapshot()
  })

  it('Buy plugins match snapshot on android + IL', () => {
    expect(collapsePlugins(buyPluginJson, 'android', 'IL')).toMatchSnapshot()
  })
})

describe('GuiPlugins tools', () => {
  it('filter android + US as expected', () => {
    const list = collapsePlugins(testJson, 'android', 'US')

    expect(titles(list)).toEqual(['Credit card', 'Wire transfer'])
    expect(pluginIds(list)).toEqual(['phony', 'gox'])
    expect(list).toMatchSnapshot()
  })

  it('filter ios + US as expected', () => {
    const list = collapsePlugins(testJson, 'ios', 'US')

    expect(titles(list)).toEqual(['Apple Pay', 'Wire transfer', 'Credit card'])
    expect(pluginIds(list)).toEqual(['phony', 'gox', 'phony'])
    expect(list).toMatchSnapshot()
  })

  it('filter ios + JP as expected', () => {
    const list = collapsePlugins(testJson, 'ios', 'JP')

    expect(titles(list)).toEqual(['Nice lawsuit'])
    expect(pluginIds(list)).toEqual(['gox'])
    expect(list).toMatchSnapshot()
  })

  it('filter everything for GB', () => {
    const list = collapsePlugins(testJson, 'ios', 'GB')

    expect(titles(list)).toEqual([])
  })
})

function pluginIds (plugins: BuySellPlugin[]): string[] {
  return plugins.map(plugin => plugin.pluginId)
}

function titles (plugins: BuySellPlugin[]): string[] {
  return plugins.map(plugin => plugin.title)
}

// This test data is designed to exercise specific problem areas:
const testJson: any = [
  {
    id: 'phony-credit',
    pluginId: 'phony',
    title: 'Credit card',
    paymentType: 'credit',
    countryCodes: { US: true, AU: true }
  },
  {
    id: 'phony-apple',
    pluginId: 'phony', // Duplicated plugin
    addOnUrl: '/applePay', // Different URL
    title: 'Apple Pay',
    paymentType: 'applepay',
    countryCodes: { US: true },
    forPlatform: 'ios'
  },
  {
    id: 'gox',
    pluginId: 'gox',
    title: 'Wire transfer',
    paymentType: {
      wire: true,
      hack: true
    },
    countryCodes: { US: true, JP: true }
  },
  '----- Sorting -----',
  {
    id: 'phony-credit',
    priority: 3
  },
  {
    id: 'gox',
    priority: 4
  },
  '----- Platform Specifics -----',
  {
    id: 'phony-apple',
    forPlatform: 'ios',
    priority: 2 // Apple Pay to top on iOS
  },
  {
    id: 'phony-credit',
    forPlatform: 'ios',
    priority: 5 // Credit to bottom on iOS
  },
  '----- Country Specifics -----',
  {
    id: 'gox',
    title: 'Nice lawsuit', // Special title in Japan
    forCountries: ['JP', 'GB'], // Note that GB isn't supported above
    priority: 1
  }
]
