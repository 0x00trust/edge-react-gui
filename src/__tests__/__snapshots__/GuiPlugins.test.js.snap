// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuiPlugins tools filter android + US as expected 1`] = `
Array [
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [],
    "description": "",
    "paymentType": "credit",
    "pluginId": "phony",
    "title": "Credit card",
  },
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [],
    "description": "",
    "paymentType": Object {
      "hack": true,
      "wire": true,
    },
    "pluginId": "gox",
    "title": "Wire transfer",
  },
]
`;

exports[`GuiPlugins tools filter ios + JP as expected 1`] = `
Array [
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [],
    "description": "",
    "paymentType": Object {
      "hack": true,
      "wire": true,
    },
    "pluginId": "gox",
    "title": "Nice lawsuit",
  },
]
`;

exports[`GuiPlugins tools filter ios + US as expected 1`] = `
Array [
  Object {
    "addOnUrl": "/applePay",
    "cryptoCodes": Array [],
    "description": "",
    "paymentType": "applepay",
    "pluginId": "phony",
    "title": "Apple Pay",
  },
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [],
    "description": "",
    "paymentType": Object {
      "hack": true,
      "wire": true,
    },
    "pluginId": "gox",
    "title": "Wire transfer",
  },
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [],
    "description": "",
    "paymentType": "credit",
    "pluginId": "phony",
    "title": "Credit card",
  },
]
`;

exports[`Production plugin data Buy plugins match snapshot on android + IL 1`] = `
Array [
  Object {
    "addOnUrl": "buy?order_id=null&page=0",
    "cryptoCodes": Array [
      "BTC",
    ],
    "description": "Fee: ~6.5% / Settlement: 1-48 hours
Limit: â‚¬5000/day",
    "partnerIconPath": "https://edge.app/wp-content/uploads/2019/08/bits-of-gold-logo-sm.png",
    "paymentType": "bank",
    "paymentTypeLogoKey": "bank",
    "pluginId": "bitsofgold",
    "title": "Israeli Bank Transfer",
  },
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [
      "BTC",
      "BCH",
      "ETH",
      "LTC",
      "XRP",
    ],
    "description": "Fee: ~7% / Settlement: 10-30 minutes
Limit: $18000/day",
    "partnerIconPath": "https://edge.app/wp-content/uploads/2019/03/simplex-logo-sm-square.png",
    "paymentType": "credit",
    "paymentTypeLogoKey": "credit",
    "pluginId": "simplex",
    "title": "Credit and Debit Card",
  },
]
`;

exports[`Production plugin data Buy plugins match snapshot on iOS + US 1`] = `
Array [
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [
      "BTC",
      "ETH",
      "DAI",
    ],
    "description": "Fee: 1% / Settlement: 1-5 days
Limit: $2500/week",
    "partnerIconPath": "https://edge.app/wp-content/uploads/2019/01/wyre-logo-square-small.png",
    "paymentType": "bank",
    "paymentTypeLogoKey": "bank",
    "pluginId": "wyre",
    "title": "ACH Bank Transfer",
  },
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [
      "BTC",
      "BCH",
      "ETH",
      "LTC",
      "XRP",
    ],
    "description": "Fee: ~7% / Settlement: 10-30 minutes
Limit: $18000/day",
    "partnerIconPath": "https://edge.app/wp-content/uploads/2019/03/simplex-logo-sm-square.png",
    "paymentType": "credit",
    "paymentTypeLogoKey": "credit",
    "pluginId": "simplex",
    "title": "Credit and Debit Card",
  },
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [
      "BTC",
      "BCH",
      "ETH",
      "DAI",
      "LTC",
      "EOS",
      "XRP",
      "XLM",
      "BAT",
      "BNB",
      "PAX",
      "TUSD",
      "USDC",
      "USDT",
    ],
    "description": "Fee: 6.5% / Settlement: 10 minutes
Limit: $2000/day",
    "partnerIconPath": "https://edge.app/wp-content/uploads/2019/05/icon_black_small.png",
    "paymentType": Object {
      "applepay": true,
      "credit": true,
    },
    "paymentTypeLogoKey": "applepay",
    "pluginId": "moonpay",
    "title": "Apple Pay and Credit Card",
  },
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [
      "BTC",
    ],
    "description": "Fee: 3-8% / Settlement: Instant
Limit: $200 - $3000/day",
    "partnerIconPath": "https://edge.app/wp-content/uploads/2019/05/libertyXlogo.png",
    "paymentType": Object {
      "cash": true,
      "credit": true,
    },
    "paymentTypeLogoKey": "cash",
    "pluginId": "libertyx",
    "title": "Cash and Debit Card",
  },
]
`;

exports[`Production plugin data Sell plugins match snapshot on iOS + US 1`] = `
Array [
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [
      "BTC",
      "ETH",
      "DAI",
    ],
    "description": "Fee: 1% / Settlement: 1-5 days
Limit: Unlimited",
    "partnerIconPath": "https://edge.app/wp-content/uploads/2019/01/wyre-logo-square-small.png",
    "paymentType": "bank",
    "paymentTypeLogoKey": "bank",
    "pluginId": "wyre",
    "title": "ACH Bank Transfer",
  },
  Object {
    "addOnUrl": "",
    "cryptoCodes": Array [
      "BTC",
      "DASH",
      "DOGE",
      "ETH",
      "LTC",
    ],
    "description": "Buy gift cards and reload phones with crypto",
    "partnerIconPath": "https://edge.app/wp-content/uploads/2019/01/bitrefill-2.png",
    "paymentType": "giftcard",
    "paymentTypeLogoKey": "giftcard",
    "pluginId": "bitrefill",
    "title": "Gift Cards",
  },
]
`;
